dbms/src/Flash/Pipeline/Schedule/Events/Event.cpp

```diff
@@ -0,0 +1,131 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Common/Exception.h>
+#include <Common/MemoryTrackerSetter.h>
+#include <Flash/Executor/PipelineExecutorStatus.h>
+#include <Flash/Pipeline/Schedule/Events/Event.h>
+#include <Flash/Pipeline/Schedule/TaskScheduler.h>
+#include <assert.h>
+
+namespace DB
+{
+// if any exception throw here, we should pass err msg and cancel the query.
+#define CATCH                                                                \
+    catch (...)                                                              \
+    {                                                                        \
+        exec_status.onErrorOccurred(getCurrentExceptionMessage(true, true)); \
+    }
+
+void Event::addInput(const EventPtr & input)
+{
+    assert(status == EventStatus::INIT);
+    input->addOutput(shared_from_this());
+    ++unfinished_inputs;
+}
+
+bool Event::withoutInput()
+{
+    assert(status == EventStatus::INIT);
+    return 0 == unfinished_inputs;
+}
+
+void Event::addOutput(const EventPtr & output)
+{
+    assert(status == EventStatus::INIT);
+    outputs.push_back(output);
+}
+
+void Event::onInputFinish()
+{
+    auto cur_value = unfinished_inputs.fetch_sub(1);
+    assert(cur_value >= 1);
+    if (1 == cur_value)
+        schedule();
+}
+
+void Event::schedule() noexcept
+{
+    switchStatus(EventStatus::INIT, EventStatus::SCHEDULED);
```

Comment:
I would change `RUNTIME_CHECK(status.compare_exchange_strong(from, to));` to `RUNTIME_ASSERT(status.compare_exchange_strong(from, to));` and then there would be no need to catch exceptions.
Because a switchStatus failure is an unintended error, there is no better way to handle it.

Reply:
updated to `RUNTIME_ASSERT`